File: Euler9.py
import numpy as np
from functools import reduce
from time import perf_counter

def EratM(N):
    """
    The Sieve of Eratosthenes implemented with numpy arrays
    Returns the list of primes lower or equal than N
    """
    if N < 11:
        first_primes = ([],[],[2],[2,3],[2,3],[2,3,5],[2,3,5],[2,3,5,7],[2,3,5,7],[2,3,5,7],[2,3,5,7])
        return(first_primes[N]) 
    N_bool = np.array([True]*(N+1))
    N_bool[0] = False
    N_bool[1] = False
    N_bool[2**2::2] = False # eiminating even numbers
    N_bool[3**2::3] = False # eiminating multiples of 3
    p = 5
    while p*p <= N:
        # start from p**2
        # because all the smaller composites have factors <p
        # and are already eliminated in previous steps
        if N_bool[p]:
            N_bool[p**2::p] = False # python doesn't care if ::p goes beyond existing array
        if N_bool[p+2]:
            N_bool[(p+2)**2::(p+2)] = False
        p = p + 6 # we can move in steps of 6
    # returns indices of nonzero elements, which in this case
    # ARE the correcponding natural numbers, which were not eliminated
    # Have to use index [0], because for technical reasons it produces a 2D array
    return list(np.nonzero(N_bool)[0]) 

def IncompletePrimeFactor(N): 
    """
    Returns the list of all prime factors of N with a possible exception
    of a single factor > sqrt(N). If N is prime, it returns an empty list.
    """
    # create the list of primes lower or equal to sqrt(N)
    primes = np.array(EratM(int(np.floor(np.sqrt(N)))))
    # if N is divisible by a prime, N%p = 0
    # so when we create an array of N%p for all primes
    # it has 0 entries for factors and non-zero one for non-factors
    # recasting it as bool turns 0 into False and non-zero to True
    # after we invert them, we get True for factors and False for non-factors
    prime_mask = np.invert(np.array(N%primes, dtype = bool))
    # now we only need to apply this mask to return only prime factors
    # but since we limited ourselves to p < sqrt(N)
    # we might miss a factor > sqrt(N), like e.g. 33 = 3*11
    # so be careful when using this output
    return list(primes[prime_mask])

def FactorMultiplicity(N):
    """
    Returns two lists, the first with all the factors of N
    the second with corresponding multimlicities of each of them.
    """
    factors = IncompletePrimeFactor(N)
    if not factors: return [N],[1] # N is prime
    multiplicity = []
    # just cycle through all the factors 
    # and check how many time N is divisible by each
    for p in factors:
        m = 0
        while N%p == 0:
            m += 1
            N = N//p
        multiplicity.append(m)
    if N != 1: # the case of a single additional factor > sqrt(N)
        factors.append(N)
        multiplicity.append(1)
        
    # must recast them as python int, because
    # since we used numpy array before, their
    # type was changed to long_scalars, the default
    # int type of numpy. If we don't do this
    # we quickly get overflow when multiplying these
    # numbers
    
    factors = [int(x) for x in factors]
    return factors, multiplicity

  
def AllDivisors(N):
    """
        Generates all divisors for a number 
        as an iterating function. Example of use:
        all_factors = [u for u in AllDivisors(20)]
        produces:
        all_factors = [1, 2, 4, 5, 10, 20]
        Note that this produces an unsorted list
        e. g. for N = 12 -> [1, 2, 4, 3, 6, 12]
    """
    factors, multiplicity = FactorMultiplicity(N) # produces lists of factors with corresponding multiplicities
    fmpairs = [(f,m) for (f,m) in zip(factors,multiplicity)] # zips them into a single list of pairs
    # e. g. for N = 60 we get [(2,2), (3,1), (5,1)]
    nfactors = len(factors)
    p = [0] * nfactors # create a list filled with zeroes for all factors
    # e. g. for N = 60  p = [0, 0, 0]
    
    # This following structure runs through 
    # all possible combinations of exponents
    # For example, for N = 60 we'll get
    # [0, 0, 0] => 1
    # [1, 0, 0] => 2
    # [2, 0, 0] => 4
    # [0, 1, 0] => 3
    # [1, 1, 0] => 6
    # [2, 1, 0] => 12
    # [0, 0, 1] => 5
    # [1, 0, 1] => 10
    # [2, 0, 1] => 20
    # [0, 1, 1] => 15
    # [1, 1, 1] => 30
    # [2, 1, 1] => 60
    
    while True: # iterator starts
        # this runs through all possible combination of factors and their powers
        # reduce with the multiplication lambda multiplies them all together
        # starts with 1, since all powers are set to 0
        # f0^p0 * f1^p1 * f2^p2 ... where each exponent p can be from 0 to m, where m is the corresponding multiplicity
        yield reduce(lambda x, y: x*y, [fmpairs[x][0]**p[x] for x in range(nfactors)], 1)
        i = 0 # after yield we always start with the first exponent
        while True:
            p[i] += 1 # increase current exponent
            if p[i] <= fmpairs[i][1]: # if it is still below or equal to the multiplicity
                break # going to yield and running reduce
            p[i] = 0 # go back to exponent = 0
            i += 1 # and shift to the next factor
            if i >= nfactors: # we have increased all exponents to max
                return # iterator stops

start = perf_counter()              
p = 1000000 # given perimeter 
k = 0
if p % 2 == 0: # for a Pythagorean triple p = 2km(m+n), so p must be even
    all_factors = [u for u in AllDivisors(p//2)] 
    if len(all_factors) != 2: # len(all_factors) == 2 means p//2 is a prime and cannot be equal to km(m+n)
        # we look for a combination of km(m+n)
        # such that 
        # m > n > 1
        # GCD(m,n) = 0
        # m and n don't have equal parity, i. e. m + n is odd
        # k is any integer
        for m in all_factors:
            if m == 1: continue
            for q in all_factors: # q = m + n
                n = q - m
                # check the three conditions
                if n >= m or n < 1: continue
                if np.gcd(m,n) > 1: continue
                if q % 2 == 0: continue
                # the resulting triple m, n, k is valid and can be used to construct a Pythagorean triangle
                k = p//(2*m*q)
                ptriple = sorted([k*m*m-k*n*n, 2*k*m*n, k*m*m+k*n*n]) # once we have k, m, n, we can generate the sides a, b, and c.
                print(ptriple, 'with sum', sum(ptriple), 'equal to', p, 'and c**2', ptriple[2]*ptriple[2], '= a**2 + b**2', ptriple[0]*ptriple[0] + ptriple[1]*ptriple[1])
                print('product a*b*c', ptriple[0]*ptriple[1]*ptriple[2])

if k == 0: print('perimeter', p, 'is impossible')

end = perf_counter()
print(end - start,'sec')


  

File: Euler9.cpp
#include <vector>
#include <iostream>
#include <cmath>
#include <algorithm>
#include <climits>

using namespace std;

/* this is basically Python script rewritten with C++ vectors etc.*/


//the sieve of Eratosthenes, the limit by both memory and time is about 10^10
// older version with step 2
// (10^10 takes about 137 seconds)
vector<unsigned long> Erat(unsigned long N){
    vector<unsigned long> primes;
    if (N == 0 || N == 1) return primes;
    primes.push_back(2);
    if (N == 2) return primes;
    primes.push_back(3);
    if (N == 3) return primes;
    vector<bool> Nbool(N+1, true);
    vector<bool>::iterator itr = Nbool.begin();
    *itr = false;
    *(itr+1) = false;
    //*(itr+2) = true;
    //*(itr+3) = true;
    for (itr = Nbool.begin() + 4; itr < Nbool.end(); itr += 2) *itr = false;
    unsigned long p = 3;
    while (p*p <= N) {
        for (itr = Nbool.begin() + p*p; itr < Nbool.end(); itr += p) *itr = false;
        p += 2;
        while (! Nbool[p]) p += 2;
    }
    for (itr = Nbool.begin()+5; itr < Nbool.end(); itr += 2) {
        if (*itr) primes.push_back(itr - Nbool.begin());
    }
    return primes;
}

//the sieve of Eratosthenes, the limit by both memory and time is about 10^10
// newer version with step 6
// (10^10 takes about 132 seconds)
vector<unsigned long> EratM(unsigned long N){
    vector<unsigned long> primes;
    if (N < 11) {
        if (N < 2) return primes;
        primes.push_back(2);
        if (N == 2) return primes;
        primes.push_back(3);
        if (N < 5) return primes;
        primes.push_back(5);
        if (N < 7) return primes;
        primes.push_back(7);
        return primes;
    }
    vector<bool> Nbool(N+1, true);
    vector<bool>::iterator itr = Nbool.begin();
    *itr = false;
    *(itr+1) = false;
    for (itr = Nbool.begin() + 4; itr < Nbool.end(); itr += 2) *itr = false;
    for (itr = Nbool.begin() + 9; itr < Nbool.end(); itr += 3) *itr = false;
    unsigned long p = 5;
    while (p*p <= N) {
        if (Nbool[p]) {
            for (itr = Nbool.begin() + p*p; itr < Nbool.end(); itr += p) *itr = false;
        }
        if (Nbool[p+2]) {
            for (itr = Nbool.begin() + (p+2)*(p+2); itr < Nbool.end(); itr += p+2) *itr = false;
        }
        p += 6;
    }
    primes.push_back(2);
    for (itr = Nbool.begin()+3; itr < Nbool.end(); itr += 2) {
        if (*itr) primes.push_back(itr - Nbool.begin());
    }
    return primes;
}


vector<unsigned long> IncompletePrimeFactor(unsigned long N) {
    vector<unsigned long> primes = EratM((unsigned long)(floor(sqrt((double) N))));
    vector<unsigned long>::iterator itr;
    vector<unsigned long> factors;
    for (itr = primes.begin(); itr != primes.end(); itr ++) {
        if (N % *itr == 0) factors.push_back(*itr);
    }
    return factors;
} 
  
vector<pair <unsigned long,int>> FactorMultiplicity(unsigned long N){
    vector<unsigned long> factors = IncompletePrimeFactor(N);
    vector<pair <unsigned long,int>> mfactors;
    if (factors.empty()) {
        mfactors.push_back(make_pair(N,1));
        return mfactors;
    }
    vector<unsigned long>::iterator itr;
    int m;
    for (itr = factors.begin(); itr != factors.end(); itr++) {
        m = 0;
        while (N % *itr == 0) {
            m++;
            N /= *itr;
        }
        mfactors.push_back(make_pair(*itr,m));
    }
    if (N != 1) mfactors.push_back(make_pair(N, 1));
    return mfactors;
}

int ipow(int base, int exp) 
// power function for integers
// works as follows:
// say we have base = 2 and exp = 5 or 101 in binary
// cycle 1: last bit is 1, base now is 2, so result becomes 2, bitshift to 10, base becomes 4
// cycle 2: last bit 0, so no change to result, shift 10 to 1, base becomes 16
// cycle 3: last bit 1, base is now 16, so result becomes 32, shift to 0, cycle terminates, returns 32
{
    int result = 1;
    for (;;) {
        if (exp & 1) // checks the last bit of exp is 1, equivalent to exp % 2
            result *= base;
        exp >>= 1; // bit shift by one bit to the right, equivalent to exp / 2
        if (!exp) // done
            break;
        base *= base;
    }
    return result;
}

unsigned long collapse(vector<pair <unsigned long,int>> mfactors, vector<int> f) 
// this is just the equivalent of python's 
// reduce(lambda x, y: x*y, [fmpairs[x][0]**p[x] for x in range(nfactors)], 1)
// i. e. take factors from the first items of pairs and exponentiate them to the given powers
{
    vector<pair <unsigned long,int>>::iterator itrmf;
    vector<int>::iterator itrf;
    unsigned long prod = 1;
    for (itrmf = mfactors.begin(), itrf = f.begin(); (itrmf != mfactors.end()) && (itrf != f.end()); itrmf++, itrf++) {
        if (*itrf != 0) prod *= ipow((*itrmf).first, *itrf);
    }
    return prod;
}

vector <unsigned long> AllDivisors(unsigned long N) {
    vector<pair <unsigned long,int>> mfactors = FactorMultiplicity(N);
    vector <unsigned long> propfactors;
    int nf = mfactors.size();
    vector<int> f(nf, 0);
    bool run = true;
    while (run) {
        propfactors.push_back(collapse(mfactors,f));
        int i = 0;
        while (true) {
            f[i] += 1;
            if (f[i] <= mfactors[i].second) break;
            f[i] = 0;
            i++;
            if (i == nf) {
                run = false;
                break;
            }
        }
    }
    return propfactors;
}

int main() {
    cout << "enter perimeter (for 10 000 000 a*b*c already overflows ulong):" << endl;
    unsigned long p;
    cin >> p;
    clock_t c_start = clock();
    int k = 0;
    unsigned long maxprod, prod;
    maxprod = 0;
    if (p % 2 == 0) {
    vector <unsigned long> factors = AllDivisors(p/2);
        if (factors.size() != 2) {
            vector <unsigned long>::iterator mm,q;
            for (mm = factors.begin(); mm != factors.end(); mm++) {
                unsigned long m = *mm;
                if (m == 1) continue;
                for (q = factors.begin(); q != factors.end(); q++) {
                    unsigned long n = *q - m;
                    if ((n >= m) || (n < 1)) continue;
                    if ((m + n) % 2 == 0) continue;
                    if (__gcd(n,m) > 1) continue;
                    k = p/(2*m**q);
                    unsigned long a = k*m*m-k*n*n;
                    unsigned long b = 2*k*n*m;
                    unsigned long c = k*m*m+k*n*n;
                    prod = a*b*c;
                    if (prod > maxprod) maxprod = prod;
                    if (a > b) swap(a,b);
                    cout << a << "\t" << b << "\t" << c << "\t" << c+b+a << "\t" << c*a*b << endl;
                } 
            }
        }
    }
    clock_t c_end = clock();
    cout << 1000.0*(c_end - c_start) / CLOCKS_PER_SEC << " ms\n";
    cout << (c_end - c_start) / CLOCKS_PER_SEC << " sec\n";
    
    if (k == 0) cout << "this perimeter is impossible!" << endl;
    //if (maxprod == 0) cout << "-1" << endl;
    else cout << "\tmaximum product: \t\t\t\t" << maxprod << endl;
    cout << "\tbeware of max unsigned long: \t\t\t" << ULONG_MAX << endl;
}

